### PR TITLE
Add SSH password authentication method

### PR DESCRIPTION
**Pull request by bOnepain**
**Description**
This feature adds a new question in the workflow, letting choose between ssh authentication with an ssh key or with a password. We often need to be able to connect to a SSH server with a password, especially on dev env with docker containers.

**Definition Of Done**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Infrastructure/Cli/Ssh.php [hunk 1]
```diff
@@ -23,9 +23,9 @@ public function __construct(string $host, int $port = 22)
         $this->port = $port;
     }
 
-    public function exec(string $command, string $username): ?string
+    public function exec(string $command, string $username, string $password = null): ?string
```
Comment (anaelChardan) [inline src/Infrastructure/Cli/Ssh.php]: Please use a nullable typehint ;)

File: src/Infrastructure/Cli/Ssh.php [hunk 2]
```diff
@@ -49,7 +49,7 @@ public function exec(string $command, string $username): ?string
         return $output;
     }
 
-    public function getAuthenticatedConnection(string $username)
+    public function getAuthenticatedConnection(string $username, $password = null)
```
Comment (anaelChardan) [inline src/Infrastructure/Cli/Ssh.php]: same here

File: src/Infrastructure/Pim/SshConnection.php [hunk 3]
```diff
@@ -24,11 +24,15 @@ class SshConnection implements PimConnection
     /** @var string */
     private $username;
 
-    public function __construct(string $host, int $port, string $username)
+    /** @var string */
+    private $password;
+
+    public function __construct(string $host, int $port, string $username, string $password = null)
```
Comment (anaelChardan) [inline src/Infrastructure/Pim/SshConnection.php]: same here ;)

File: src/Infrastructure/Pim/SshConnection.php [hunk 4]
```diff
@@ -56,4 +60,9 @@ public function getUsername(): string
     {
         return $this->username;
     }
+
+    public function getPassword(): string
```
Comment (anaelChardan) [inline src/Infrastructure/Pim/SshConnection.php]: same here ;)

**Review by anaelChardan — COMMENTED** [review]
Thanks a lot! Could you please have a look on "https://github.com/akeneo/transporteo/blob/master/tests/spec/Infrastructure/MigrationStep/S010FromReadyToSourcePimConfiguredSpec.php" for the tests please ;)

**Comment by bOnepain** [comment]
Hello,
I've added the nullable typehints as you required. Concerning the tests, I'm afraid I do not master php spec enough to edit the code properly. I must confess I do not understand everything in this spec file.

**Comment by gplanchat** [comment]
@anaelChardan I'm looking the tests to implement, but there is something where I'm a bit lost.

In the tests, ont the `SshConnection` instanciation there is a 4th argument of type `SshKey`:
* https://github.com/akeneo/transporteo/blob/master/tests/spec/Infrastructure/MigrationStep/S010FromReadyToSourcePimConfiguredSpec.php#L149
* https://github.com/akeneo/transporteo/blob/master/tests/spec/Infrastructure/MigrationStep/S010FromReadyToSourcePimConfiguredSpec.php#L247 

The thing is, the `SshConnection` constructor only has 3 arguments: https://github.com/akeneo/transporteo/blob/master/src/Infrastructure/Pim/SshConnection.php#L27

~What is the correct behavior here ?~ @bOnepain answered the question :)

**Comment by anaelChardan** [comment]
Indeed, you're right ;) If it is possible for @bOnepain it would be good to fix at the same time this problem ;). Thanks for everything ;)

**Comment by bOnepain** [comment]
I'm on it

**Comment by bOnepain** [comment]
I've added some phpspecs. Tell me if it's what was expected or if I've missed something...

File: src/Infrastructure/Pim/SshConnection.php [hunk 5]
```diff
@@ -24,11 +24,15 @@ class SshConnection implements PimConnection
...
+
+    public function __construct(string $host, int $port, string $username, ?string $password = null)
     {
         $this->host = $host;
         $this->port = $port;
         $this->username = $username;
+        $this->password = $password;
     }
 
     public static function fromString(string $serverInformation)
```
Comment (anaelChardan) [inline src/Infrastructure/Pim/SshConnection.php]: maybe add the password here if it exists :

```
$parsedServerInformation['password'] ?? null

**Review by anaelChardan — APPROVED** [review]

**Comment by anaelChardan** [comment]
Thanks a lot! ;)

**Pull request merged**